File: DATN-Fashion-Shop-BE/src/main/java/com/example/DATN_Fashion_Shop_BE/model/Product.java
package com.example.DATN_Fashion_Shop_BE.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.envers.Audited;
import org.hibernate.envers.NotAudited;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
@Entity
@Table(name = "products")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Audited
public class Product extends BaseEntity{
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "status", length = 255, nullable = false)
    private String status;

    @Column(name = "base_price", nullable = false)
    private Double basePrice;

    @Column(name = "is_active", nullable = false)
    private Boolean isActive;

    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
            name = "products_categories",
            joinColumns = @JoinColumn(name = "product_id"),
            inverseJoinColumns = @JoinColumn(name = "category_id")
    )
    private List<Category> categories = new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @NotAudited
    private List<ProductsTranslation> translations;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @NotAudited
    private List<ProductVariant> variants;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @NotAudited
    private List<ProductMedia> medias;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @NotAudited
    private List<Review> Review;

    @ManyToOne
    @JoinColumn(name = "promotion_id", nullable = true)
    private Promotion promotion;

    @ManyToOne
    @JoinColumn(name = "review_id", nullable = true)
    @NotAudited
    private Review review;

    public ProductsTranslation getTranslationByLanguage(String langCode) {
        ProductsTranslation translation = translations.stream()
                .filter(t -> t.getLanguage().getCode().equals(langCode)
                        && t.getName() != null && !t.getName().isEmpty())
                .findFirst()
                .orElse(null);

        // Nếu không tìm thấy, trả về bản dịch với mã ngôn ngữ "en"
        if (translation == null) {
            translation = translations.stream()
                    .filter(t -> t.getLanguage().getCode().equals("en"))
                    .findFirst()
                    .orElse(null);
        }

        return translation;
    }

}

File: FE/DATN-Fashion-Shop-FE/src/app/models/Product/product.ts
import { BaseEntity} from '../BaseEntity';
import {Category, CategoryImpl} from '../Category/Category';
import { ProductsTranslation} from './ProductsTranslation';
import { ProductVariant} from './ProductVariant';
import { Promotion} from './Promotion';

export class Product extends BaseEntity {
  id: number;
  status: string;
  basePrice: number;
  isActive: boolean;
  categories: CategoryImpl[];
  translations: ProductsTranslation[];
  variants: ProductVariant[];
  promotion?: Promotion;

  constructor(data?: Partial<Product>) {
    super(data);
    this.id = data?.id || 0;
    this.status = data?.status || '';
    this.basePrice = data?.basePrice || 0;
    this.isActive = data?.isActive ?? true;
    this.categories = data?.categories?.map(c => new CategoryImpl(c)) || [];
    this.translations = data?.translations?.map(t => new ProductsTranslation(t)) || [];
    this.variants = data?.variants?.map(v => new ProductVariant(v)) || [];
    this.promotion = data?.promotion ? new Promotion(data.promotion) : undefined;
  }

  getTranslationByLanguage(langCode: string): ProductsTranslation | undefined {
    let translation = this.translations.find(
      t => t.language?.code === langCode && t.name?.trim()
    );

    if (!translation) {
      translation = this.translations.find(t => t.language?.code === 'en');
    }

    return translation;
  }
}
